### ISSUE TITLE
Prevent backwards incompatible spec changes via SAT & DAT

### ISSUE THREAD

**Issue by sherifnada** [airbytehq/airbyte#14097]
## Tell us about the problem you're trying to solve
Today, it is entirely possible via developer error to create a new version of a connector which doesn't work with configs accepted by previous versions of the connector. For example, apart from human vigilance, there is nothing preventing me from marking a previously optional spec field as `required`, or completely renaming a field in a way that breaks all existing customer configs. 

Today, we don't have config migrations (see #5396) which means it's not possible to automatically converge all connector configs to a new format. 

We also don't have a mechanism to alert users to this in the UI or elsewhere as described in #2516. 

Absent both of these things, we need to prevent backwards incompatible changes to specs. Otherwise there is nothing preventing

## Describe the solution you’d like
In SAT & DAT, add a test which pulls the `:latest` image for a connector, compares its spec against the spec being tested, and fails if there is a backwards incompatible change (if there is no `:latest` image then assume this is the first publication of this connector). For custom connector developers (not working within the airbyte monorepo), we may need to disable this test occasionally, so it's probably useful to have a flag that can disable this. 

## Sub-items

- [x] #14222
  - make sure to consult notes below!

Spec validation
- [ ] #14593
- [ ] #14595
- [ ] DAT: Validate that connector spec schema changes are backwards-compatible
- [ ] DAT: Validate that configs generated with previous spec are valid according to new spec
- [x] #14600
- [ ] DAT: Assert that `additionalProperties: false` is not specified in connector spec schema

Catalog validation
- [ ] SAT: Validate that connector catalog changes are backwards-compatible

**Comment by pedroslopez** [comment]
Related P0 postmortem caused by this: https://docs.google.com/document/d/1ru9ApRI7fGtv5gNECchsQe6EB20LlY4UZhJOtEHvOAw/edit#

**Comment by sherifnada** [comment]
backlog grooming notes: 
- the main challenge will be in producing a config using the previous schema, that works on the current spec. 
  - we could auto-generate this? 
  - compare the specs directly? 
  - actually have an old config and run tests with the config. We could do this by uploading all configs that were used to test & publish version N, then 
- maybe just generally enforce `additionalProperties: true` for all objects in SAT & DAT
- Should backwards compatibility be syntactic (i.e: JSON schemas should work together) or semantic (Previous configs should pass future versions' `CHECK` method)? 
- Basically we feel like everything is a ~hack~ compromise until we have config migrations
- Actually - this builds into config migrations. If a connector fails the machinery we have here then there _must_ be a config migration for it. 

sorta converging on: 
phase 0: syntactic equivalence between configs
phase 1: semantic equivalence (an actual physical config from previous versions must pass future versions)
phase 2: add config migrations

**Comment by alafanechere** [comment]
I'm closing this as the RFC was written [here](https://www.notion.so/Preventing-Backwards-Incompatible-Changes-in-Connectors-e05dafd0f1a349dd819b569a8a4564a7) and I created an [EPIC](https://github.com/airbytehq/airbyte/issues/15290) for the first phase of the project.
